Feature segmentation fastDRaW

hgueziri opened the pull request:
## Description

Hi all,

I implemented a new segmentation feature, the FastDRaW, which is based on the random walker algorithm to perform image segmentation.

The basic idea is to perform random walker segmentation in two steps. The first step is applied on a small (down-sampled) version of the image resulting in a coarse segmentation result. The second step is to refine the result by applying a second random walker segmentation over a narrow strip around the coarse result. In addition, a region of interest (ROI) is extracted during labeling (drawing seeds). The role of this ROI is to reduce the computation time and reduce the required number of labels to draw.

Main benefits are:
1- A important gain in the computation time (almost 10x faster)
2- FastDRaW introduces a new feature for ROI extraction which reduces the number of labels required to achieve the segmentation

Since the approach is mainly dedicated to interactive segmentation, I am preparing an Introductory example to be added in the gallery examples section.

This is my first contribution to scikit-image, please let me know for any modifications and/or explanatory details that can clarify the implementation.

Here are some comparative segmentation examples. Original image (top), Random Walker algorithm (middle) and FastDRaW (bottom):

![ex_1](https://cloud.githubusercontent.com/assets/8172629/17943245/95a04bc4-6a08-11e6-851b-0a2d5be5e71e.png) ![ex_4](https://cloud.githubusercontent.com/assets/8172629/17943247/95a684a8-6a08-11e6-9b4b-a5dd613c2c17.png) ![ex_5](https://cloud.githubusercontent.com/assets/8172629/17943246/95a38dca-6a08-11e6-9c22-1200b29f6ec3.png)
## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Unit tests
## References

This branch implemented the method following:
H.-E. Gueziri, L. Lakhdar, M. J. McGuffin and C. Laporte, "FastDRaW – Fast Delineation by Random Walker: application to large images", [accepted for publication] in MICCAI workshop on Interactive Medical Image Computing, Athens, Greece (2016).

jni commented:
Hi @hgueziri!

Thanks for your PR!

My first instinct, though, is that this method is just too new to come into scikit-image. Other core members should chime in, but in my view the idea behind scikit-image is to provide established algorithms. We don't have a hard citation limit like scikit-learn does, but something that hasn't even been published yet is pretty risky!

As such, I would recommend that you put this in your own repo, and make an announcement on our mailing list — I'm sure many would be interested in trying it out! Then, if a bunch of people find it useful, we can revisit it.

As to the code itself, I had a quick look, and my major piece of advice is that you dramatically shorten the functions — you should aim for around 25 lines of code in a function. This isn't a hard limit, but a guideline to ensure that code is broadly readable and understandable.

I'll leave a couple of smaller comments on the code itself.

Also, if you haven't published to PyPI before, it's pretty easy! Have a look at the official guide here:
https://packaging.python.org/distributing/
That will make it really easy for others to find and use this method.

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+bar" to run both segmentation. The computation time is displayed.
+
+.. [1] H.-E. Gueziri, L. Lakhdar, M. J. McGuffin and C. Laporte,
+    *FastDRaW - Fast Delineation by Random Walker: application to large
```
Indent (align with "H" on first line), remove bold.

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+        yy = int(event.ydata)
+        xx = int(event.xdata)
+        # draw green pixels on the rgb image
+        rgb_image[yy-brush:yy+brush+1,xx-brush:xx+brush+1] = (0,255,0)
```
Use PEP8 spacing, as you did above

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+"""
+# drawing callback function
+def onDraw(event):
+    """This function is called when the mouse moves over the image
```
Put triple quotes on same line for one-liner docstrings

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+def onKeypress(event):
+    """This function is called when keyboard buttons are pressed
+    """
+    global target_label
```
Is there any way around using a global here?

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+        # --- display the contour in red over the original image
+        contour = (dilation(segm, disk(1)) - segm).astype(np.bool)
+        result = original_image.copy()
+        result[contour,:] = (255,0,0)
```
PEP8 spacing, here and elsewhere

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+    
+
+"""
+Main
```
What are "callback functions" and "main"? In plaintext parts, try to address novice coders, who might not have heard these terms before. e.g. "First we write functions to define what needs to happen when the user clicks on the image", "Now we write the main analysis script that makes use of those functions."

jni commented on doc/examples/segmentation/plot_fast_draw_segmentation.py:
```diff
@@ -0,0 +1,198 @@
+ax3 = fig.add_subplot(133)
+ax3.imshow(original_image)
+ax3.set_title('RW')
+plt.axis('off')
```
Have a look at our other examples. We've mostly standardised around `fig, axes = plt.subplots(1, 3)`. And I think `axes[0, 2].axis('off')` works, but I might be wrong there.

jni commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,278 @@
+from scipy.sparse import csr_matrix, coo_matrix
+from scipy.sparse.linalg import cg as cg_solver
+
+from ..transform import resize as skresize
```
Don't alias here, since you're not using any other resize function...

jni commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,278 @@
+        self.image = rgb2grey(image)
+        ## It is important to normalize the data between [0,1] so that beta
+        ## makes sens
+        self.image = self.image / 255.0
```
Use `skimage.img_as_float`

jni replied:
(which is nice because it doesn't rescale it if the image is already a float in [0, 1])

jni commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,278 @@
+        and returns the graph's Laplacian `L`.
+        """
+        # Building the graph: vertices, edges and weights
+        nsize = reduce(lambda x,y: x*y,image.shape,1)
```
PEP8 again

jni commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,278 @@
+        
+        gr_right = np.abs(image[:, :-1] - image[:, 1:]).ravel()
+        gr_down = np.abs(image[:-1] - image[1:]).ravel()
+        weights = np.exp(- self.beta * np.r_[gr_right, gr_down]**2)+1e-6
```
Unary -: no space before operand.

jni replied:
ie `-self.beta`, not `- self.beta`

jni commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,278 @@
+        ds_proba[(ds_labels == target_label) & (self.ds_maskROI)] = 1   
+        
+        # 3.4- Compute the corse segmentation result and the refinement region
+        #      around the corse result
```
As mentioned in my general comment, this function is way too long. Your section titles are good candidates for where to break it up into component functions.

jni commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,278 @@
+                ds_entropyMap +=  distMap
+        
+        # 2.3- Normalize the energy map and compute the ROI
+        ds_entropyMap = ds_entropyMap / ds_entropyMap.max()
```
PEP8: `ds_entropy_map`

jni commented:
@hgueziri I've left a few code comments, I hope they are useful whether or not we end up merging!

hgueziri commented:
Hi @jni,

Thank you for taking the time to read the code. 

I addressed all the issues you highlighted in the comments. As you mentioned, this will help to improve the readability of the code even for a private repo. So, thanks a lot.

I understand that it might be too soon to add the method to scikit-image. I just thought that it could be interesting to add an "interactive" method to the toolbox, such as the example using a matplotlib window for drawing on the image to perform the segmentation (which is the native idea behind random walker / graph cuts algorithms).

Any other suggestion would be welcome.

On a different note, do you have any idea why the Travis-CI test keeps failing for the example given in the docstring?

JDWarner commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,336 @@
+        and returns the graph's Laplacian `L`.
+        """
+        # Building the graph: vertices, edges and weights
+        nsize = reduce(lambda x, y : x*y, image.shape,1)
```
This is the cause of the test problem, `reduce` is not imported or defined. I'm surprised it gets to the doctests - this should fail in the usual test suite.

emmanuelle commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,336 @@
+    def __init__(self, image, beta=300, downsampled_size=100,
+                 tol=1.e-3, return_full_prob=False):
+        
+        assert (beta > 0), 'beta should be positive.'
```
Here it would be better to raise a ValueError than an AssertionError

emmanuelle commented on skimage/segmentation/fast_draw.py:
```diff
@@ -0,0 +1,336 @@
+                         shape=(pixel_nb, pixel_nb))
+        L = lap.tocsr()
+        
+        return L
```
Couldn't you use the function ``_make_laplacian_sparse`` from the ``random_walker_segmentation.py`` module? The less code the better :-).

emmanuelle commented:
Hi @hgueziri thank you for your interest in scikit-image! The algorithm looks very interesting, I think I will try it on my own images at some point (it would need to be extended to 3-D images, though -- but it shouldn't be too difficult I think?). Oh, and congratulations for the MICCAI paper :-).

I agree with @jni that the algorithm is a bit too recent (not published yet from what I understand?) to be integrated into scikit-image. However, if you're interested, I can suggest that we keep on reviewing your PR until it is "scikit-image compatible", so that if the algorithm grows popular we can merge it in a while. What do you think?

I haven't reviewed the PR in detail, but one thing that I noticed is that you use object-oriented programming, with one class for the algorithm. This programming workflow is usually very convenient at the beginning, but it is less modular on the long run. For example, in the ``random_walker_segmentation.py`` module, the algorithm is coded with functions only, so that it would be possible to re-use some of these functions in other modules, if necessary (like the function to build the Laplacian, as I suggested in a comment). It is less convenient if you don't have functions, but methods of a class. Even for your future research projects it might be useful for you to be able to re-use some functions (which lowers the probability of bugs ;-)).

So, even if it involves passing parameters (which shouldn't have any impact on performance) from one function to the other, I'd recommend that you write the algorithm using functions instead of a class, for the sake of modularity and maintainability. What do you think?

hgueziri commented:
Hi @emmanuelle, 
Thank you for your support and review. As I mentioned, I'll be happy to adapt the PR so that it satisfies scikit-image compatibility requirements, for an eventual merging.

- I am working on my free time to extend the code to the 3D images, so it may take few time though.
- The `_make_laplacian_sparse` function was inspired from the `random_walker_segmentation.py` code. It could be easily replaced by the one from the RW. I'll work on it for the next commit.
- I completely agree with you that functional programming will be easier to be re-used elsewhere. The main reason why I choose object oriented programming is that I (maybe) "over" simplified the code to gain computation time. Object oriented programming allows to avoid creating the graph's Laplacian matrices (`L` and `ds_L`) and resize the image `ds_image` at each segmentation call. Since the approach is meant to be used in interactive settings (segmentation refresh every ~ 0.1 s), this could be computationally cumbersome. However, I'll perform some tests to have a practical comparison between the two.

Best,

stefanv commented:
@hgueziri Thank you for your interest in contributing to scikit-image.  I agree with Juan and Emmanuelle's assessment that this method is too new to include in scikit-image, but I think it would be great to have this out there.  I'd suggest that you a) put this in a git repo and b) write a blog post about it.  We'd be happy to help you disseminate such a post through our social network.

I'm going to close this for now (we have way too many issues open!), but let's look into it again in the future, especially once 3D has been implemented.

Thanks again, and best regards.
